### ISSUE TITLE
problems with 2D Probe diagnostics in 3D simulation

### ISSUE THREAD

**Issue by jpsikal** [SmileiPIC/Smilei#206]
Dear SMILEI developers,

I report my problems when using Probe diagnostics in 3D simulation. 
It seems that there is a bug in this diagnostics. 

I used Field diagnostics and Probe diagnostics and I imaged electron densities in x-y (with constant z in the center) and x-z (with constant y in the center) planes. 

I found a good agreement in the x-y plane for both diagnostics, but I found some problems in x-z plane with Probe diagnostics.

Electron density in x-y plane is attached here (the same picture can be seen for Probe and Field diagnostics in this plane and for Field diagnostics in x-z plane)
![Ne-xy-Probe](https://user-images.githubusercontent.com/53087545/73830764-8153a380-4805-11ea-9f02-7d5032bf7e88.png)
 
Electron density in x-z plane from Probe diagnostics is attached here (unexpected profile of density can be seen - this profile differ from Field diagnostics data)
![Ne-xz-Probe](https://user-images.githubusercontent.com/53087545/73831055-fc1cbe80-4805-11ea-856d-29f383523ece.png)

I used Smilei version 4.3

The simulation is large (containing about 16e+9 particles). 

Namelist is copied here (some parts with other diagnostics were removed to make it shorter):

```
import math
import numpy 
l0 = 2.0*math.pi               # laser wavelength [in code units]
t0 = l0                        # optical cycle
Lsimx = 128.*l0               # length of the simulation box in x (= 102.4 mum)
Lsimy = 44.8*l0               # length of the simulation box in y (= 35.84 mum) 
Lsimz = 44.8*l0               # length of the simulation box in z (= 35.84 mum)
Tsim = 4.01*t0
resx = 25.
resy = 25.
resz = 25.                     # nb of cells on laser wavelength
rest = resx/0.9                # nb of timesteps in one optical cycle
dt = t0/rest/math.sqrt(3.0) 

ne = 4.5                   # electron density (code units => 1 = plasma critical density)
Te = 0.00587              # Te normalised in mec^2 (code units) (= 3000 eV)
nppc = 4                  # max. number of particle-per-cell
mp = 1836.                # proton mass
mC = 22032.               # carbon mass

FWHMinI  = 3.75*l0  #(= 3 mum)
waistinI = FWHMinI/(2.0*math.sqrt(math.log(2.0)))
waistinE = FWHMinI/math.sqrt(2.0*math.log(2.0))

FWHMtinI = 11.25*t0                  #(= 30 fs)
FWHMtinE = FWHMtinI*math.sqrt(2.0)  

diagEvery   = int(3.0*t0/dt)    # frequency of outputs for DiagField
diagStart = int(1.0*t0/dt)

nx = 128*25            # nb of cells along x-axis in the simulation box
ny = int(44.8*25.)
nz = int(44.8*25.)

Main(
    geometry = "3Dcartesian",
    interpolation_order = 4,    
    timestep = dt,
    simulation_time = Tsim,
    cell_length = [l0/resx,l0/resy,l0/resz],
    grid_length  = [Lsimx,Lsimy,Lsimz],    
    number_of_patches = [128,32,32],    # one patch = 25x35x35 cells
    EM_boundary_conditions = [
        ["silver-muller","silver-muller"],
        ["silver-muller","silver-muller"],
        ["silver-muller","silver-muller"]
    ], 
    print_every = 5,   
    random_seed = 0
)

Species(
    name = 'electron',
    position_initialization = 'random',
    momentum_initialization = 'maxwell-juettner',
    particles_per_cell = gaussian(nppc, xvacuum=18.75*l0, xlength=62.5*l0, 
        xfwhm=44.2*l0, xcenter=50.*l0, xorder=2, yorder=0, zorder=0),
    mass = 1.0,
    charge = -1.0,
    number_density = gaussian(ne, xvacuum=18.75*l0, xlength=62.5*l0, 
        xfwhm=26.01*l0, xcenter=50.*l0, xorder=2, yorder=0, zorder=0),
    temperature = [Te],
    thermal_boundary_temperature = [Te],
    boundary_conditions = [ ['remove'], ['thermalize'], ['thermalize'] ],
    time_frozen = 3.5*t0
)

Species(
    name = 'proton',
    position_initialization = 'random',
    momentum_initialization = 'maxwell-juettner',
    particles_per_cell = gaussian(nppc, xvacuum=18.75*l0, xlength=62.5*l0, 
        xfwhm=44.2*l0, xcenter=50.*l0, xorder=2, yorder=0, zorder=0),
    mass = mp, 
    charge = 1.0,
    number_density = gaussian(ne/4.0, xvacuum=18.75*l0, xlength=62.5*l0, 
        xfwhm=26.01*l0, xcenter=50.*l0, xorder=2, yorder=0, zorder=0),
    temperature = [Te],
    thermal_boundary_temperature = [Te],
    boundary_conditions = [ ['remove'], ['thermalize'], ['thermalize'] ],
    time_frozen = 3.5*t0
)

Species(
    name = 'carbon',
    position_initialization = 'random',
    momentum_initialization = 'maxwell-juettner',
    particles_per_cell = gaussian(nppc, xvacuum=18.75*l0, xlength=62.5*l0, 
        xfwhm=44.2*l0, xcenter=50.*l0, xorder=2, yorder=0, zorder=0),
    mass = mC, 
    charge = 6.0,
    number_density = gaussian(ne/8.0, xvacuum=18.75*l0, xlength=62.5*l0, 
        xfwhm=26.01*l0, xcenter=50.*l0, xorder=2, yorder=0, zorder=0),
    temperature = [Te],
    thermal_boundary_temperature = [Te],
    boundary_conditions = [ ['remove'], ['thermalize'], ['thermalize'] ],
    time_frozen = 3.5*t0
)

LaserGaussian3D(
    box_side        = "xmin",
    a0              = 52.94,  # intensity 6.0e+21
    omega           = 1.,
    focus           = [37.5*l0, Main.grid_length[1]/2., Main.grid_length[2]/2.],
    waist           = waistinE,
    incidence_angle = [0.,0.],
    polarization_phi = 0.,
    ellipticity     = 0.,
    time_envelope   = tcosine(base=0., start=0., duration=12.*t0, 
        phi=math.pi/2., freq=1./(12.0*2.0))
)


LoadBalancing(
    initial_balance = True,
    every = 50
)

### DIAGNOSTICS

DiagFields(
    every = [diagStart,diagEvery],
    fields = ['Rho_electron'],
    subgrid = numpy.s_[::2, ::2, ::2]
)

list_fields = ['Ex','Ey','Bz','Rho_electron','Rho_proton','Rho_carbon']

# probe in x-y plane
DiagProbe(
    every = [diagStart,diagEvery],
    origin = [0., 0., Main.grid_length[2]/2.],
    corners = [
        [Main.grid_length[0], 0., Main.grid_length[2]/2.],
        [0., Main.grid_length[1], Main.grid_length[2]/2.]
    ],
    number = [nx, ny],
    fields = list_fields
)

# probe in x-z plane
DiagProbe(
    every = [diagStart,diagEvery],
    origin = [0., Main.grid_length[1]/2., 0.],
    corners = [
        [Main.grid_length[0], Main.grid_length[1]/2., 0.],
        [0., Main.grid_length[1]/2., Main.grid_length[2]]
    ],
    number = [nx, nz],
    fields = list_fields
)
```

If you need I can provide more info (scripts for simulation run on computer cluster, scripts for post-processing of data etc.)

Thanks in advance for you help,

Jan

**Comment by mccoys** [comment]
Damn. I have never seen this behaviour before and this should not happen. I will investigate asap.

It might be only a problem of post processing. Probe points need to be ordered and this might cause some strange behaviour.

**Comment by mccoys** [comment]
If this is possible for you, could you try the same diagnostic with 4 times less points in each direction? I am just thinking that large probes are treated slightly differently.

**Comment by jpsikal** [comment]
So, I have tried the same diagnostic with 4 times less points in each direction (see the namelist): 
....
....
nx = 128*25/4     
ny = int(44.8*25./4)
nz = int(44.8*25./4)
....
....
....
list_fields = ['Rho_electron']

# probe in x-y plane
DiagProbe(
    every = [diagStartP,diagEveryP],
    origin = [0., 0., Main.grid_length[2]/2.],
    corners = [
        [Main.grid_length[0], 0., Main.grid_length[2]/2.],
        [0., Main.grid_length[1], Main.grid_length[2]/2.]
    ],
    number = [nx, ny],
    fields = list_fields
)

# probe in x-z plane
DiagProbe(
    every = [diagStartP,diagEveryP],
    origin = [0., Main.grid_length[1]/2., 0.],
    corners = [
        [Main.grid_length[0], Main.grid_length[1]/2., 0.],
        [0., Main.grid_length[1]/2., Main.grid_length[2]]
    ],
    number = [nx, nz],
    fields = list_fields
)

....
....


Electron density in x-y plane is the same as in previous case. 

Electron density in x-z plane changed: 
![Ne-xzprobe](https://user-images.githubusercontent.com/53087545/73935792-aa496680-48e1-11ea-9496-c981676fa0be.png)

So, it seems there is a bug in the calculation of data values on a defined grid for Probe diagnostics.

**Comment by mccoys** [comment]
Nevermind my previous message. I managed to reproduce the bug and I am trying to see where it comes from. There is definitely a problem when ordering the points.

**Comment by mccoys** [comment]
After some investigations, I figured that the problem is not directly in probes, but in the fourth-order interpolator (there is a specific version that only probes use). The second order seems to work. I will now try to find that error in the interpolator.

**Comment by jpsikal** [comment]
O.k., thanks for help. I need to use fourth-order interpolation and Probe diagnostics since I plan to run large 3D simulation. I would need smaller cells when using two-order interpolation, that is almost impossible for my parameters. Field diagnostics is slow and produce a huge amount of data, so I use it only in large time interval and with subgrids. So, I would be happy if you let me know what to modify in the source code when you will find the error.

**Comment by mccoys** [comment]
I found the error. I will try to test and push the corrections tomorrow

**Comment by mccoys** [comment]
Please try the last push and close the issue if fixed for you.

**Comment by jpsikal** [comment]
The problem was solved as you modified three lines in src/Interpolator/Interpolar3D4Order.cpp. 
Thanks a lot! Jan

**Comment by vasimos** [comment]
Hello! 

I have a similar problem in 3D simulations trying to use 2D probe diagnostics, despite using the latest version from (github v4.3-106-g8c22fd7-master, downloaded 2020-03-26). I only tried a x-y plane diagnostic and as can be seen in the figure I get some 'empty' patches in the density. I attach my input file. Please let me know of any other information needed.

![density_xy_probe](https://user-images.githubusercontent.com/2751008/75427745-58778780-5947-11ea-8a7c-0692593291e2.png)

[input.txt](https://github.com/SmileiPIC/Smilei/files/4260534/input.txt)

**Comment by beck-llr** [comment]
This is actually a different problem. It is not directly linked to the probes since it has sometimes been observed in fields too.  It is usually random (blank patches are different from one simulation to another) and we are not able to reproduce it. It seems to depend on the file system, the hdf5 version, mpi version and so on ...

**Comment by MickaelGrech** [comment]
@vasimos can you tell me how many MPI and openMP threads you use for this simulation?
I will try to run it on Irene.

**Comment by vasimos** [comment]
@MickaelGrech I attach my slurm script. Thanks.

[job.slurm.txt](https://github.com/SmileiPIC/Smilei/files/4261130/job.slurm.txt)

**Comment by jpsikal** [comment]
I can confirm what was written by Arnaud Beck. I observed "blank" patches using Fields diagnostics. Example is attached. In my case, it was solved only by using different computer cluster ... but I did not try to vary hdf5 and mpi version on the same cluster. 
![Ey120](https://user-images.githubusercontent.com/53087545/75558798-a8874480-5a42-11ea-8689-164ba1d3da0d.png)

Nevertheless , the reason for that problem is different than the issue I reported 23 days ago.
